File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseKGroup(head *ListNode, k int) *ListNode {
    if head == nil || k == 1 {
        return head
    }
    
    dummy := &ListNode{Val: 0, Next: head}
    prev := dummy
    curr := head
    var next *ListNode
    
    count := 0
    for temp := head; temp != nil; temp = temp.Next {
        count++
    }
    
    for count >= k {
        curr = prev.Next
        next = curr.Next
        for i := 1; i < k; i++ {
            curr.Next = next.Next
            next.Next = prev.Next
            prev.Next = next
            next = curr.Next
        }
        prev = curr
        count -= k
    }
    
    return dummy.Next
}

File: 25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        if (head == null || k == 1) return head;
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode curr = head;
        ListNode next = null;
        
        int count = 0;
        for (ListNode temp = head; temp != null; temp = temp.next)
            count++;
        
        while (count >= k) {
            curr = prev.next;
            next = curr.next;
            for (int i = 1; i < k; i++) {
                curr.next = next.next;
                next.next = prev.next;
                prev.next = next;
                next = curr.next;
            }
            prev = curr;
            count -= k;
        }
        
        return dummy.next;
    }
}